pkg/build/controller/common/util_test.go

```diff
@@ -0,0 +1,157 @@
+	}
+	if build.Status.CompletionTimestamp == nil {
+		t.Errorf("should have set the CompletionTimestamp, but instead it was nil")
+	}
+	if build.Status.Duration > 0 {
+		t.Errorf("should have set the Duration to 0s, but instead it was %v", build.Status.Duration)
+	}
+}
+
+func TestHandleBuildPruning(t *testing.T) {
+	var objects []runtime.Object
+	buildconfig, builds := mockBuildsList(10)
+
+	objects = append(objects, &buildconfig)
+	for index := range builds {
+		objects = append(objects, &builds[index])
+	}
+
+	osclient := testclient.NewSimpleFake(objects...)
+
+	build, err := osclient.Builds("namespace").Get("myapp-0", metav1.GetOptions{})
+	if err != nil {
+		t.Errorf("%v", err)
+	}
+
+	buildLister := buildclient.NewOSClientBuildClient(osclient)
+	buildConfigGetter := buildclient.NewOSClientBuildConfigClient(osclient)
+	buildDeleter := buildclient.NewBuildDeleter(osclient)
+
+	bcName := buildutil.ConfigNameForBuild(build)
+	successfulStartingBuilds, err := buildutil.BuildConfigBuilds(buildLister, build.Namespace, bcName, func(build buildapi.Build) bool { return build.Status.Phase == buildapi.BuildPhaseComplete })
+	sort.Sort(ByCreationTimestamp(successfulStartingBuilds.Items))
+
+	failedStartingBuilds, err := buildutil.BuildConfigBuilds(buildLister, build.Namespace, bcName, func(build buildapi.Build) bool { return build.Status.Phase == buildapi.BuildPhaseFailed })
+	sort.Sort(ByCreationTimestamp(failedStartingBuilds.Items))
+
+	if len(successfulStartingBuilds.Items)+len(failedStartingBuilds.Items) != 10 {
+		t.Errorf("should start with 10 builds, but started with %v instead", len(successfulStartingBuilds.Items)+len(failedStartingBuilds.Items))
+	}
+
+	if err := HandleBuildPruning(bcName, build.Namespace, buildLister, buildConfigGetter, buildDeleter); err != nil {
+		t.Errorf("error pruning builds: %v", err)
+	}
+
+	successfulRemainingBuilds, err := buildutil.BuildConfigBuilds(buildLister, build.Namespace, bcName, func(build buildapi.Build) bool { return build.Status.Phase == buildapi.BuildPhaseComplete })
+	sort.Sort(ByCreationTimestamp(successfulRemainingBuilds.Items))
+
+	failedRemainingBuilds, err := buildutil.BuildConfigBuilds(buildLister, build.Namespace, bcName, func(build buildapi.Build) bool { return build.Status.Phase == buildapi.BuildPhaseFailed })
+	sort.Sort(ByCreationTimestamp(failedRemainingBuilds.Items))
+
+	if len(successfulRemainingBuilds.Items)+len(failedRemainingBuilds.Items) != 5 {
+		t.Errorf("there should only be 5 builds left, but instead there are %v", len(successfulRemainingBuilds.Items)+len(failedRemainingBuilds.Items))
+	}
+
+	if !reflect.DeepEqual(successfulStartingBuilds.Items[:2], successfulRemainingBuilds.Items) {
+		t.Errorf("the two most recent successful builds should be left, but they were not")
+	}
+
+	if !reflect.DeepEqual(failedStartingBuilds.Items[:3], failedRemainingBuilds.Items) {
+		t.Errorf("the two most recent failed builds should be left, but they were not")
```

Comment:
s/two/three/

and for both this and the case above, it's generally useful to print the actual and expected values (in this case the lists of builds) to help debug if it does fail.